iree/compiler/Conversion/LinalgToVector/VectorizeLinalgOps.cpp

```diff
@@ -0,0 +1,339 @@
+    return success();
+  }
+};
+
+template <typename OpTy>
+struct VectorizeCmpOp : public OpConversionPattern<OpTy> {
+  using OpConversionPattern<OpTy>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      OpTy op, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    if (llvm::any_of(args,
+                     [](Value arg) { return arg.getType().isIntOrFloat(); })) {
+      return failure();
+    }
+    auto newOp =
+        rewriter.create<OpTy>(op.getLoc(), op.predicate(), args[0], args[1]);
+    rewriter.replaceOp(op, newOp.getResult());
+    return success();
+  }
+};
+
+struct VectorizeSelectOp : public OpConversionPattern<mlir::SelectOp> {
+  using OpConversionPattern<mlir::SelectOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      mlir::SelectOp op, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    if (llvm::any_of(args,
+                     [](Value arg) { return arg.getType().isIntOrFloat(); })) {
+      return failure();
+    }
+    auto newOp =
+        rewriter.create<SelectOp>(op.getLoc(), args[0], args[1], args[2]);
+    rewriter.replaceOp(op, newOp.getResult());
+    return success();
+  }
+};
+
+struct VectorizeYieldOp : public OpConversionPattern<linalg::YieldOp> {
+  using OpConversionPattern<linalg::YieldOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      linalg::YieldOp yieldOp, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    if (!llvm::all_of(args,
+                      [](Value v) { return v.getType().isa<VectorType>(); })) {
+      return failure();
+    }
+    SmallVector<Value, 4> vecArgs;
+    for (Value arg : args) {
+      Value val = legalizeToVectorType(rewriter, arg);
+      if (!val) return failure();
+      vecArgs.push_back(val);
+    }
+    rewriter.replaceOpWithNewOp<linalg::YieldOp>(yieldOp, vecArgs);
+    return success();
+  }
+};
+
+struct VectorizeGenericOp : public OpConversionPattern<linalg::GenericOp> {
+  using OpConversionPattern<linalg::GenericOp>::OpConversionPattern;
+  LogicalResult matchAndRewrite(
+      linalg::GenericOp genericOp, ArrayRef<Value> args,
+      ConversionPatternRewriter &rewriter) const override {
+    // Ignore a case that either a nop pattern or a linalg.fill pattern.
+    if (isa<linalg::YieldOp>(
+            genericOp.region().front().getOperations().front())) {
+      return failure();
+    }
+
+    if (llvm::any_of(genericOp.iterator_types(), [](Attribute attr) {
+          return attr.cast<StringAttr>().getValue() !=
+                 getParallelIteratorTypeName();
+        })) {
+      return failure();
+    }
+
+    // Do not vectorize if one of the operand is 0-D or one of the operand is
+    // not iterated on contiguous memory.
+    for (auto map : genericOp.getIndexingMaps()) {
+      if (map.getNumResults() == 0) return failure();
+      int lastDimPos =
```

Comment:
You can just check this way I think

```
AffineDimExpr innerMostExpr = map.getResult().back().dyn_cast<AffineDimExpr>();
if (!innerMostExpr || innerMostExpr.getPosition() != map.getNumDims() - 1) {
  return failure();
}
```

Reply:
I think we want to handle the case that not iterates on innermost loop. Eg,

```
for i = ...
  for j = ...
     %0 = load %arg0[i] : memref<100xf32>
     %1 = load %arg1[i, j] : memref<100x100xf32>
     ...
```

something could be from broadcast_add.

I just found that this is not the case in our E2E tests. But I think it is still valid if we broadcast to different dim.

In Linalg form, can be something like

```
func @broadcast_add() {
  %0 = iree.placeholder for "interface buffer" {binding = @legacy_io::@arg0} : memref<4xf32>
  %1 = iree.placeholder for "interface buffer" {binding = @legacy_io::@arg1} : memref<4x4xf32>
  %2 = iree.placeholder for "interface buffer" {binding = @legacy_io::@ret0} : memref<4x4xf32>
  linalg.generic {args_in = 2 : i64,
                  args_out = 1 : i64,
                  indexing_maps = [affine_map<(d0, d1) -> (d0)>,         <---------- here's the case
                                   affine_map<(d0, d1) -> (d0, d1)>,
                                   affine_map<(d0, d1) -> (d0, d1)>],
                  iterator_types = ["parallel", "parallel"]
  } %0, %1, %2 {
  ^bb0(%arg0: f32, %arg1: f32, %arg2: f32):  // no predecessors
    %3 = addf %arg0, %arg1 : f32
    linalg.yield %3 : f32
  }: memref<4xf32>, memref<4x4xf32>, memref<4x4xf32>
  return
}
```

We can also relax the condition later if you think it's better. What do you think?

Reply:
After thinking a while, I think we can relax the condition later.